src/IBM.Cloud.SDK.Core/Service/IBMService.cs

```diff
@@ -71,272 +56,41 @@ protected string Endpoint
-                apikey: !string.IsNullOrEmpty(ApiKey) ? ApiKey : null,
-                iamUrl: !string.IsNullOrEmpty(IamUrl) ? IamUrl : null,
-                userManagedAccessToken: !string.IsNullOrEmpty(IamAccessToken) ? IamAccessToken : null,
-                iamClientId: !string.IsNullOrEmpty(ClientId) ? ClientId : null,
-                iamClientSecret: !string.IsNullOrEmpty(ClientSecret) ? ClientSecret : null
-                );
-
-            BasicAuthConfig basicAuthConfig = new BasicAuthConfig(
-                username: !string.IsNullOrEmpty(UserName) ? UserName : null,
-                password: !string.IsNullOrEmpty(Password) ? Password : null
-                );
-
-            if (!string.IsNullOrEmpty(ServiceUrl))
-                Endpoint = ServiceUrl;
-
-            if (!string.IsNullOrEmpty(Icp4dAccessToken) && !string.IsNullOrEmpty(Icp4dUrl))
-            {
-                SetAuthenticator(icp4dConfig);
-            }
-
-            if (!string.IsNullOrEmpty(ApiKey))
-            {
-                SetAuthenticator(iamConfig);
-            }
-
-            if (!string.IsNullOrEmpty(UserName) && !string.IsNullOrEmpty(Password))
-            {
-                SetAuthenticator(basicAuthConfig);
-            }
-        }
-
-        protected IBMService(string serviceName, string url)
-        {
-            ServiceName = serviceName;
-            Client = new IBMHttpClient(url, UserName, Password);
-
-            if (!string.IsNullOrEmpty(Endpoint))
-                Endpoint = url;
-            //TODO: verificar como iremos obter de um arquivo json por injeção de dependencia
-            //ApiKey = CredentialUtils.GetApiKey(serviceName);
-            //Endpoint = CredentialUtils.GetApiUrl(serviceName);
+            Client = new IBMHttpClient();
+            authenticator = ConfigBasedAuthenticatorFactory.GetAuthenticator(serviceName);
+            var creds = CredentialUtils.GetServiceProperties(serviceName);
+            creds.TryGetValue("URL", out string url);
+            SetEndpoint(string.IsNullOrEmpty(url) ? defaultEndpoint : url);
         }
 
         protected IBMService(string serviceName, string url, IClient httpClient)
         {
-            ServiceName = serviceName;
-            Client = httpClient;
+            ServiceName = serviceName;
+            Client = httpClient ?? throw new ArgumentNullException(nameof(httpClient));
+            authenticator = new NoauthAuthenticator();
 
             if (!string.IsNullOrEmpty(Endpoint))
                 Endpoint = url;
-
-            //TODO: verificar como iremos obter de um arquivo json por injeção de dependencia
-            //ApiKey = CredentialUtils.GetApiKey(serviceName);
-            //Endpoint = CredentialUtils.GetApiUrl(serviceName);
         }
 
-        protected IBMService(string serviceName, IAuthenticatorConfig authenticatorConfig)
+        protected IBMService(string serviceName, Authenticator authenticator)

```

Comment:
This probably doesn't matter much in the .NET SDK, but...  I think it would be better to use IAuthenticator for the datatype of the parameter since that is the interface.    Someone could theoretically implement an IAuthenticator while NOT extending Authenticator, if that makes sense.   This is all theoretical though :)

Reply:
Using IAuthenticator now when possible